### ISSUE TITLE
Complexity porting to music?

### ISSUE THREAD

**Issue by Saltb0xApps** [declare-lab/TangoFlux#8]
Hey! I was wondering how difficult would it be to port this to work with text to music? 
Would it be simply retraining on a large dataset of music files and captions (instead of SFX), or there might be additional improvements needed?

**Comment by soujanyaporia** [comment]
Hi! Yes we observed simply training it on the music data can generate great music! Pls do try and tell us about the experience. Thanks !!

**Comment by Saltb0xApps** [comment]
@soujanyaporia That is great news! I'm currently trying out a training run on around 10k audio files, will let you know how it goes!

What would you recommend to be the best strategy to implement generating longer than 30s audio?
I'm considering implementing text tokens + 10s of audio conditioning during training, and during inference, pass the 10s of the previous chunk to generate an additional 20s, allowing effectively infinite duration.

Also, do you plan on open sourcing the GUI used for DPO training as shown in the original paper?
<img width="962" alt="Screenshot 2025-01-05 at 1 59 10 PM" src="https://github.com/user-attachments/assets/d0ce2689-3dcb-40dd-956f-5510785b38a1" />
While fairly easy to create from scratch, it would make it easy to replicate the exact training environment!

**Comment by soujanyaporia** [comment]
That sounds great! 

DPO training: this UI was used for human evaluation. the DPO training data was automatically created using CLAP scores.

**Comment by Saltb0xApps** [comment]
Ah, understood. So basically the steps to train a model are -
1. train a base model for ~100 epochs
2. Generate outputs from base model and rank outputs with CLAP scores automatically
3. Assemble CLAP results into preference pairs
4. DPO finetuning with these preference pairs to further improve quality

**Comment by soujanyaporia** [comment]
That is right. You may fine-tune TangoFlux-base on your data and then do CRPO. Looking forward to trying your model :)

**Comment by Saltb0xApps** [comment]
Absolutely! Before starting a large scale training, I'm currently working on figuring out an ideal way to extend past 30s of audio while maintaining consistency, which is essential for music generation

Have you run any tests with higher number of trainable parameters than 515M, if that yields better audio quality?

**Comment by soujanyaporia** [comment]
We are now training a large model :) More details will be shared soon. With music prompt-based fine-tuning, you should be able to extend the audio beyond 30 seconds. 

In our preliminary experiments, we found it generated excellent music. Bu

**Comment by Saltb0xApps** [comment]
https://github.com/user-attachments/assets/38c97600-5526-44c9-8c73-a51080f8efff

Seems to be a good start!

Config: Epoch 16, 5k audio files, 5e-4 lr (constant with warmup schedule), batch size 12. 
Prompt: "hip hop beat with drums"

**Comment by soujanyaporia** [comment]
This sounds so cool! Once you finish training, it will be great to upload the model to HF (if you wish) and create an HF space.

**Comment by Saltb0xApps** [comment]
@

**Comment by Saltb0xApps** [comment]
@soujanyaporia do you believe this config is comparable to your successful pre-training run? I feel like the grad norm is a bit aggressive. Batch size is 48, lr is 5e-5, 4xA100 80GB GPUs, 60 epochs, 325k files.

https://wandb.ai/akhiltolani/Text%20to%20Audio%20Flow%20matching/runs/9htvvnl1/workspace?nw=nwuserakhiltolani


I had a few issues with exploding gradients in the previous runs -
https://wandb.ai/akhiltolani/Text%20to%20Audio%20Flow%20matching?nw=nwuserakhiltolani

**Comment by soujanyaporia** [comment]
Yeah, I think it is a case of exploding gradient. @hungchiayu1 could you please share the pre-training config we used?

**Comment by Saltb0xApps** [comment]
@soujanyaporia it seems like the issue stems from my modifications of the model and training pipeline for audio conditioning (to allow longer than 30s creations), not the pre-training config itself.

Do you believe this approach is correct for extending creations past 30s?
https://github.com/declare-lab/TangoFlux/compare/main...Saltb0xApps:TangoFlux:main

**Comment by soujanyaporia** [comment]
I did not have time to read the function carefully, but for more than 30 seconds of generation, I would fine-tune the model by conditioning on previously generated samples.

**Comment by Saltb0xApps** [comment]
@soujanyaporia after training on a music dataset for a few days, the _quality_ of the output music has significantly improved, but seems to be stuck on a plateau harmonically.

The first 10 seconds are prefixed music, and the remaining 20 seconds after that is generated by the model for continuation -

https://github.com/user-attachments/assets/b0411668-95f5-40b6-927a-5dce2265ad51

While the "quality" of the remaining 20 seconds is really good, its not harmonic. 


I wanted to ask that during your experiments on music datasets, were you able to generate fully coherent music?

If not, do you believe the limitation is on stable audio's VAE, or this is just a case of requiring a lot more training? 

Any guidance is greatly appreciated! :)

**Comment by soujanyaporia** [comment]
Thanks for the updates. I think the model needs more training or data to adapt to the audio prompts. The base 10 seconds seem to be very good suggesting the model is capable of generating good music. In our tests, we observed excellent music performance as well.

**Comment by soujanyaporia** [comment]
One option could be to first fine-tune on music data and then fine-tune it again to extend the music beyond 30 seconds. What do you think?

**Comment by Saltb0xApps** [comment]
My assumption is that with the finetuning approach, the outputs wont exactly have a seamless transition between chunks. If we try to generate a 3 minute long track, it wont really be consistent over that duration, and the model would change overall style of the track every 30s, leaving the last chunk very different than the first chunk, and not really a complete consistent track.

I have continuation/outpainting working logically by fusing 10s of audio with the encoder hidden states, but thats resulting in a very big sequence length of 1332 
```
fused_hidden_states = self.input_embedding(
    latents_input, cond, encoder_hidden_states, duration_hidden_states
)
fused_hidden_states shape: torch.Size([1, 1332, 64])

latents (shape [1,645,64])
cond (also [1,645,64])
fused text+duration embedding ([1,42,64])
total: 645 + 645 + 42 = 1332
```

compared to the default model's 42 sequence length
```
encoder_hidden_states = torch.cat(
      [encoder_hidden_states, duration_hidden_states], dim=1
  )  ## (bs,seq_len,dim)
encoder_hidden_states shape: torch.Size([1, 42, 1024])
```

I'm guessing this is the core reason why the model is having a hard time generating coherent music with continuation. I'm exploring alternative strategies like cross attention or just simply reducing the sequence length to be closer to the original model's sequence length.

**Comment by Saltb0xApps** [comment]
For reference, here's the current state of music with more training. It seems to be harmonically improving over epochs. Initial 10s is fixed from the training dataset.

https://github.com/user-attachments/assets/498ec2a2-669c-4fa0-8cda-585417139ca1

**Comment by soujanyaporia** [comment]
Yeah. This sounds better

**Comment by Saltb0xApps** [comment]
@soujanyaporia quick question - is it better to have more detailed captions like 
```
"Epic orchestral welcome theme in D Major, 120 BPM, featuring powerful drums, percussion build-up, and sweeping strings for a grand atmosphere."
```

or is it better to keep captions generic
```
Epic orchestral welcome theme
```

**Comment by soujanyaporia** [comment]
It really depends on the data. But a robustly trained model should work well on both these types of captions.

**Comment by WangHaoyuuu** [comment]
Hello! I’m also trying to fine-tune a model on a music dataset like MusicCaps. However, whether I load from a checkpoint or not, the first 10 seconds of the generated music always have the same tone. Have you encountered the same issue?

**Comment by hungchiayu1** [comment]
I did try to finetune on music, but I did not encounter this issue. Can i check what is your dataset size, finetuning hyperparam ?

**Comment by WangHaoyuuu** [comment]
Thank you for your response.

I'm currently working with two datasets:

Musicaps: ~5,000 music clips (10s each)

MusicBench: ~50,000 music clips (10s each)

Another question: Are [TangoFlux](https://huggingface.co/declare-lab/TangoFlux) and [TangoFlux-base](https://huggingface.co/declare-lab/TangoFlux-base) the same model? which should i resume or load?

Here are my training-related parameters:
Training-related parameters  
training:  
  per_device_batch_size: 4  
  learning_rate: 5e-4  
  gradient_accumulation_steps: 1  
  num_train_epochs: 10  
  num_warmup_steps: 1000  
  max_audio_duration: 30  

trained on 2 x A800 

Additionally, I’m encountering an issue ([#28](https://github.com/declare-lab/TangoFlux/issues/28)) where setting cfg > 1 results in completely noisy music generation.

Would appreciate any guidance on these!

**Comment by hungchiayu1** [comment]
Nope TangoFlux and TangoFlux-base are different model. TangoFlux undergoes alignment from the TangoFlux-base checkpoint. When i finetune on music, I use the TangoFlux-base model instead. Could you try finetuning with larger batch size and more epochs? My global batch size was 48 iirc and 40 epochs.
 I will get back to you regarding the other issue soon

**Comment by WangHaoyuuu** [comment]
I noticed Tangoflux max_audio_duration is set to 30 seconds, but since my clips are 10 seconds long, I’ve adjusted it to 10.
